source/game/kart/KartMove.cc

```diff
@@ -1593,6 +1639,139 @@ void KartMove::landTrick() {
     activateBoost(KartBoost::Type::TrickAndZipper, duration);
 }
 
+/// @addr{0x8058498C}
+void KartMove::enterCannon() {
+    init(true, true);
+    physics()->clearDecayingRot();
+    m_boost.resetActive();
+    state()->setBoost(false);
+
+    cancelJumpPad();
+    m_rampBoost = 0;
+    state()->setRampBoost(false);
+    m_ssmtCharge = 0;
+    m_ssmtLeewayTimer = 0;
+    m_ssmtDisableAccelTimer = 0;
+    m_bSsmtCharged = false;
+    m_bSsmtLeeway = false;
+    m_offroadInvincibility = 0;
+    state()->setBoostOffroadInvincibility(false);
+
+    dynamics()->reset();
+
+    clearDrift();
+    state()->setHop(false);
+    state()->setInCannon(true);
+    state()->setSkipWheelCalc(true);
+    state()->setCannonStart(false);
+
+    const auto [cannonPos, cannonDir] = getCannonPosRot();
+    m_cannonEntryPos = pos();
+    m_cannonEntryOfs = cannonPos - pos();
+    m_cannonEntryOfsLength = m_cannonEntryOfs.normalise();
+    m_cannonEntryOfs.normalise();
+    m_dir = m_cannonEntryOfs;
+    m_vel1Dir = m_cannonEntryOfs;
+    m_cannonOrthog = EGG::Vector3f::ey.perpInPlane(m_cannonEntryOfs, true);
+    m_cannonProgress.setZero();
+}
+
+/// @addr{0x80584D58}
+void KartMove::calcCannon() {
+    auto [cannonPos, cannonDir] = getCannonPosRot();
+    EGG::Vector3f forwardXZ = cannonPos - m_cannonEntryPos - m_cannonProgress;
+    EGG::Vector3f forward = forwardXZ;
+    f32 forwardLength = forward.normalise();
+    forwardXZ.y = 0;
+    forwardXZ.normalise();
+    EGG::Vector3f local94 = m_cannonEntryOfs;
+    local94.y = 0;
+    local94.normalise();
+    m_speedRatioCapped = 1.0f;
+    EGG::Matrix34f cannonOrientation;
+    cannonOrientation.makeOrthonormalBasis(forward, EGG::Vector3f::ey);
+    EGG::Vector3f up = cannonOrientation.multVector33(EGG::Vector3f::ey);
+    m_smoothedUp = up;
+    m_up = up;
+
+    if (forwardLength < 30.0f || local94.dot(forwardXZ) <= 0.0f) {
+        exitCannon();
+        return;
+    }
+    cannonOrientation.mulRow33(1, EGG::Vector3f::ez);
+    m_speed = m_baseSpeed;
+    const auto *cannonPoint =
+            System::CourseMap::Instance()->getCannonPoint(state()->cannonPointId());
+    s16 cannonParameterIdx = cannonPoint->parameterIdx();
+    if (cannonParameterIdx < 1) {
```

Comment:
it hates me and wants me to do `s16 cannonParameterIdx = std::max<s16>(0, cannonPoint->parameterIdx());`. it otherwise infers that it's supposed to be`std::max<int>`, and gets mad when i provide s16 cnpt->parameteridx. alternative is 
`s16 cannonParameterIdx = std::max((short)0, cannonPoint->parameterIdx());`
or 
`s16 cannonParameterIdx = std::max(static_cast<short>(0), cannonPoint->parameterIdx());`

Reply:
In this case, `cannonParameterIdx` should be `size_t` to avoid the following static_cast.

Reply:
```cpp
    size_t cannonParameterIdx = std::max<size_t>(0uz, cannonPoint->parameterIdx());
    assert(cannonParameterIdx < CANNON_PARAMETERS.size());
```
like this?